### PR TITLE
Kkraune/test details

### PR DESCRIPTION
**Pull request by kkraune**
rewrote the whole thing as I learned the concepts ...

will use this for https://docs.vespa.ai/documentation/getting-started-vespa-applications.html - and update basic-search-java/README.md as well, for a better intro how to write and test searchers

will also add more javadoc - this is first use of Searcher and the likes for new users

### CODE REVIEW

File: basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        if (!tree.isEmpty()) {
+            Item rootItem = tree.getRoot();
+            if (isMetalQuery(rootItem)) {
+                OrItem orItem = new OrItem();
+                orItem.addItem(rootItem);
+                orItem.addItem(new WordItem("metal", "album"));
+                tree.setRoot(orItem);
+            }
+        }
+        query.trace("Metal added", true, 2);
```
Comment (bratseth) [inline basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java]: Move the trace into the if block.

File: basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    public MetalSearcher() {
+        metalWords = new ArrayList<>();
+    }
+
+    @Override
+    public Result search(Query query, Execution execution) {
+        Model model = query.getModel();
+        QueryTree tree = model.getQueryTree();
+        if (!tree.isEmpty()) {
```
Comment (bratseth) [inline basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java]: i suggest you move this condition into the isMetalQuery item so this gets easier to read.

Comment (bratseth) [reply basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java]: Then you could also invert it and do if (tree.isEmpty()) return false;

File: basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        metalWords = config.metalWords();
+    }
+
+    public MetalSearcher() {
+        metalWords = new ArrayList<>();
+    }
+
+    @Override
+    public Result search(Query query, Execution execution) {
+        Model model = query.getModel();
```
Comment (bratseth) [inline basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java]: Unnecessary variable, just dot multiple times.

File: basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import com.yahoo.prelude.query.Item;
+import com.yahoo.prelude.query.OrItem;
+import com.yahoo.prelude.query.WordItem;
+import com.yahoo.search.Query;
+import com.yahoo.search.Result;
+import com.yahoo.search.Searcher;
+import com.yahoo.search.query.Model;
+import com.yahoo.search.query.QueryTree;
+import com.yahoo.search.searchchain.Execution;
+import com.yahoo.yolean.chain.After;
```
Comment (bratseth) [inline basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java]: Unused import ... but maybe you should use it.

File: basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java (hunk 2 above)
Comment (kkraune) [reply basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java]: thanks. the searcher was also wrong, fixed

File: basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java (hunk 4 above)
Comment (kkraune) [reply basic-search-java/src/main/java/com/mydomain/example/MetalSearcher.java]: yes, used to - moved the after to services.xml but can put both places just for the example

**Pull request merged**